post/index.go

```diff
@@ -0,0 +1,369 @@
+									JOIN posts_author_links pa ON pa.post_id = $1 AND pa.author_id = a.id`, post.Id)
+	if err != nil {
+		log.Error("Error while fetching author: ", err)
+		c.AbortWithStatus(http.StatusInternalServerError)
+		return
+	}
+
+	err = repository.Db.Get(&post.Author.Image, `SELECT f.id, f.name, f.alternative_text, f.caption, f.width, f.height, f.formats as format, f.hash, f.ext, f.mime, 
+												f.size, f.url, f.preview_url, f.provider, f.provider_metadata, f.folder_path, f.created_at, f.updated_at 
+												FROM files f 
+												JOIN files_related_morphs frm ON frm.related_id = $1 AND frm.related_type = 'api::author.author' AND frm.file_id = f.id`, post.Author.Id)
+	if err != nil {
+		log.Error("Error while fetching author image: ", err)
+		c.AbortWithStatus(http.StatusInternalServerError)
+		return
+	}
+
+	var authorFormats Formats
+
+	if err := json.Unmarshal([]byte(post.Author.Image.Format), &authorFormats); err != nil {
+		log.Error("Error while unmarshal author image formats: ", err)
+		c.AbortWithStatus(http.StatusInternalServerError)
+		return
+	}
+
+	post.Author.Image.Format = ""
+	post.Author.Image.Formats = authorFormats
+
+	err = repository.Db.Get(&post.Cta, `SELECT c.id, c.name, c.component_name, c.created_at, c.updated_at 
+									FROM ctas c
+									JOIN posts_cta_links pc ON pc.post_id = $1 AND pc.cta_id = c.id`, post.Id)
+	if err != nil {
+		if err == sql.ErrNoRows {
+			log.Info("No CTA found")
+			post.Cta = Cta{}
+		} else {
+			log.Error("Error while fetching cta: ", err)
+			c.AbortWithStatus(http.StatusInternalServerError)
+			return
+		}
+	}
+
+	if post.IsResource {
+		posts := []RecommendedPost{}
+		err = repository.Db.Select(&posts, `SELECT id, title, content, slug, published_on, is_featured, created_at, updated_at, published_at,
+									summary, blog_content, meta_description, toc, tags as tag, is_published, keywords, new_content, new_toc, new_blog_content,
+									is_resource, reading_time 
+									FROM posts 
+									WHERE slug != $1`, slug)
+		if err != nil {
+			log.Error("Error while fetching recommended post: ", err)
+			c.AbortWithStatus(http.StatusInternalServerError)
+			return
+		}
+
+		post.RecommendedPosts = repository.FilterPostsByTags(posts, repository.ExtractTagNames(post.Tags))
+	}
+
+	c.JSON(http.StatusOK, post)
+}
+
+// Assuming you have a function to filter posts based on tags
+func (repository *Repository) FilterPostsByTags(posts []RecommendedPost, tags []string) []RecommendedPost {
+	// Filter posts based on tags
+	// You should replace this with your filtering logic
+	filteredPosts := []RecommendedPost{}
+	count := 0
+
+	for _, post := range posts {
+		if count >= 3 {
+			break
+		}
+
+		var postTags []Tag
+
+		if err := json.Unmarshal([]byte(post.Tag), &postTags); err != nil {
+			log.Error("Error while unmarshal tags: ", err)
+			return nil
+		}
+
+		post.Tags = postTags
+
+		for _, tag := range post.Tags {
+			if utils.Contains(tags, tag.Name) {
```

Comment:
How about creating one common function that fetches posts from given Ids, that can be used at multiple places? Here, we can collect post_ids from matched tags and then fetch posts by given post_ids.